### PR TITLE
Add value pools non finalized state

### PR DESCRIPTION
**Pull request by oxarbitrage**
## Motivation

As part of the value pools design we need to add value pools to the non finalized state.

### Specifications

https://zebra.zfnd.org/dev/rfcs/0012-value-pools.html#consensus-rules

### Designs

https://zebra.zfnd.org/dev/rfcs/0012-value-pools.html

## Solution

Start the implementation of value pools in the Non Finalized state part of the codebase; this is a draft that attempts to show the places where we need to code. 

## Review

If you have the time @teor2345 and want to take a look, i will keep working on it.

A bunch of tests fails, mostly by constraint violations, we need to look careful at the signs. Some others fail by not having the correct utxos.

### Reviewer Checklist

  - [ ] Code implements Specs and Designs
  - [ ] Tests for Expected Behaviour
  - [ ] Tests for Errors

---
This change is [Reviewable](https://reviewable.io/reviews/zcashfoundation/zebra/2610)

### CODE REVIEW

File: zebra-state/src/service/non_finalized_state/chain.rs [hunk 1]
```diff
@@ -429,6 +443,13 @@ impl UpdateWith<ContextuallyValidBlock> for Chain {
             self.update_chain_state_with(sapling_shielded_data_per_spend_anchor)?;
             self.update_chain_state_with(sapling_shielded_data_shared_anchor)?;
             self.update_chain_state_with(orchard_shielded_data)?;
+
+            // add the value balance data
+            self.update_chain_state_with(
+                &transaction
+                    .value_balance(new_outputs)
```
Comment (teor2345) [inline zebra-state/src/service/non_finalized_state/chain.rs:450]: `new_outputs` contains the UTXOs *created* by the block.
But the value balance needs the UTXOs *spent* by the block.

Spent UTXOs can come from earlier transactions in the block, the parent non-finalized chain, and the finalized state.

`check::utxo::transparent_spend` returns all the UTXOs spent by a block:
https://github.com/ZcashFoundation/zebra/blob/a176c499ab9062cd88e003f4d193903f0f68e8d0/zebra-state/src/service/non_finalized_state.rs#L182-L188

**Review by teor2345 — CHANGES_REQUESTED** [review]
Can you write down what this PR needs to do, step by step?
Then we can work out any details that are unclear.

The list should look like this list - but it can be shorter:
https://github.com/ZcashFoundation/zebra/pull/2599#pullrequestreview-728080172

I think you'll need a bunch of changes from PR #2599 for the code and tests.
And we'll need the *spent* UTXOs to get the block value balance.

**Comment by oxarbitrage** [comment]
A draft pull request is intended to ask questions, share stuff and will never be merged as it is blocked. Requesting changes in draft pull requests is aggressive and should be avoided.

**Comment by teor2345** [comment]
> A draft pull request is intended to ask questions, share stuff and will never be merged as it is blocked. Requesting changes in draft pull requests is aggressive and should be avoided.

Sorry, I didn't realise this is how change requests on draft PRs come across.
Do you want to let the team know at the retrospective?

**Comment by oxarbitrage** [comment]
I never saw anyone in the team blocking a draft PR.

**Comment by oxarbitrage** [comment]
Except you of course that do it all the time :)

**Pull request closed without merging**